exec/Chainweb/Gaps.hs

```diff
@@ -5,155 +5,250 @@
 
 module Chainweb.Gaps ( gaps ) where
 
+import           Chainweb.Api.BlockHeader
 import           Chainweb.Api.ChainId (ChainId(..))
-import           Chainweb.Api.Common (BlockHeight)
 import           Chainweb.Api.NodeInfo
 import           Chainweb.Database
 import           Chainweb.Env
 import           Chainweb.Lookups
-import           Chainweb.Worker (writeBlock)
+import           Chainweb.Worker (writeBlocks)
 import           ChainwebDb.Types.Block
 import           ChainwebData.Genesis
 import           ChainwebData.Types
-import           Control.Arrow
 import           Control.Concurrent
 import           Control.Concurrent.Async
-import           Control.Scheduler (Comp(..), traverseConcurrently_)
+import           Control.Concurrent.STM
+import           Control.Monad
+import           Control.Monad.Catch
+import           Control.Scheduler
+import           Data.Bool
 import           Data.ByteString.Lazy (ByteString)
-import           Data.Foldable
 import           Data.IORef
 import           Data.Int
-import qualified Data.IntSet as S
-import           Data.List.NonEmpty (NonEmpty(..))
-import qualified Data.List.NonEmpty as NEL
-import           Data.Maybe
+import qualified Data.Map.Strict as M
 import qualified Data.Pool as P
-import           Data.Semigroup
 import           Data.String
 import           Data.Text (Text)
+import           Data.Vector (Vector)
+import qualified Data.Vector as V
 import           Database.Beam hiding (insert)
 import           Database.Beam.Postgres
+import           Database.PostgreSQL.Simple
+import           Database.PostgreSQL.Simple.Types
+import           GHC.Natural (Natural)
 import           System.Logger hiding (logg)
+import           System.Exit (exitFailure)
 import           Text.Printf
 
 ---
 
-gaps :: Env -> Maybe Int -> IO ()
-gaps env delay = do
+gaps :: Env -> GapArgs -> IO ()
+gaps env args = do
   ecut <- queryCut env
   case ecut of
     Left e -> do
       let logg = _env_logger env
       logg Error "Error querying cut"
       logg Info $ fromString $ show e
-    Right cutBS -> gapsCut env delay cutBS
+    Right cutBS -> gapsCut env args cutBS
 
-gapsCut :: Env -> Maybe Int -> ByteString -> IO ()
-gapsCut env delay cutBS = do
+-- TODO Make this configurable?
```

Comment:
Yeah, most of the constants we're using should probably be configurable with reasonable defaults.